src/ReinforcementLearningZoo/src/algorithms/dqns/NFQ.jl

```diff
@@ -0,0 +1,87 @@
+export NFQ
+
+using Flux
+using Functors: @functor
+
+"""
+    NFQ{A<:AbstractApproximator, F, R} <: AbstractLearner
+    NFQ(approximator::A, num_iterations::Integer epochs::Integer, loss_function::F, batch_size::Integer, rng::R, γ::Float32) where {A<:AbstractApproximator, F, R}
+Neural Fitted Q-iteration as implemented in [1]
+
+# Keyword arguments
+- `approximator::AbstractApproximator` neural network
+- `num_iterations::Integer` number of value iteration iterations in FQI loop (i.e. the outer loop)
+- `epochs` number of epochs to train neural network per iteration
+- `loss_function::F` loss function of the NN
+- `sampler::BatchSampler{SARTS}` data sampler
+- `rng::R` random number generator
+- `γ::Float32` discount rate
+
+# References
+[1] Riedmiller, M. (2005). Neural Fitted Q Iteration – First Experiences with a Data Efficient Neural Reinforcement Learning Method. In: Gama, J., Camacho, R., Brazdil, P.B., Jorge, A.M., Torgo, L. (eds) Machine Learning: ECML 2005. ECML 2005. Lecture Notes in Computer Science(), vol 3720. Springer, Berlin, Heidelberg. https://doi.org/10.1007/11564096_32
+"""
+Base.@kwdef struct NFQ{A, R} <: AbstractLearner
+    approximator::A
+    num_iterations::Integer = 20
+    epochs::Integer = 100
+    loss_function::Any = mse
+    rng::R = Random.GLOBAL_RNG
+    γ::Float32 = 0.9f0
+end
+
+function NFQ(;
+    approximator::A,
+    num_iterations::Integer = 20,
+    epochs::Integer = 1000,
+    loss_function::Any = mse,
+    rng=Random.GLOBAL_RNG,
+    γ::Float32 = 0.9f0,
+    ) where {A}
+    NFQ(approximator, num_iterations, epochs, loss_function, rng, γ)
+end
+
+# Copied from BasicDQN but sure whether it's appropriate
+@functor NFQ (approximator,)
+
+RLCore.forward(L::NFQ, s::AbstractArray) = RLCore.forward(L.approximator, s)
+
+# Flux.functor(x::NFQ) = (Q = x.approximator,), y -> begin
+#     x = @set x.approximator = y.Q
+#     x
+# end
+
+function RLCore.forward(learner::NFQ, env::AbstractEnv)
+    as = action_space(env)
+    return vcat(repeat(state(env), inner=(1, length(as))), transpose(as)) |> x -> send_to_device(device(learner.approximator), x) |> x->RLCore.forward(learner, x) |> send_to_host |> vec 
+end
+
+# Avoid optimisation in the middle of an episode
+function RLBase.optimise!(::NFQ, ::NamedTuple) end
+
+# Instead do optimisation at the end of an episode
+function Base.push!(agent::Agent{<:QBasedPolicy{<:NFQ}}, ::PostEpisodeStage, env::AbstractEnv)
+    for batch in agent.trajectory
+        _optimise!(agent.policy.learner, batch, env)
+    end
+end
+
```

Comment:
@HenriDeh @jeremiahpslewis the main branch seems to have the same issue in `RLCore.optimise!(::AbstractPolicy, ::S, ::Trajectory) where {S<:AbstractStage}`. `optimise!(policy, stage, batch)` is never called in the PostEpisode stage except at the very end of the simulation.

Reply:
Do you mean at the end of the Experiment ? Are you sure it is not due to the `threshold` or the `ratio` parameters of the Trajectory ?

Reply:
I am not too familiar with the Trajectory syntax. I've created a [separate branch](https://github.com/CasBex/ReinforcementLearning.jl/tree/debugNFQ) to show what I mean. If you test the NFQ example, only the `PostActStage` is actually called

Reply:
So the issue was that:
```
function RLBase.optimise!(policy::AbstractPolicy, stage::S, trajectory::Trajectory) where {S<:AbstractStage}
    for batch in trajectory
        optimise!(policy, stage, batch)
    end
end
``` 
in base.jl is dispatched on every stage and not only the stages where you want to optimise. This means that at every post-act-stage, your policy would sample the trajectory and call the dummy optimise of the stage. Once at the end of the episode, because the ratio of the controller is 1, the trajectory returns nothing when looped, because your already sampled it once per step (which is what the ratio = 1 wants). 

@jeremiahpslewis this is significant bug caught here. 

@CasBex I'm taking the liberty of pushing a fix to your branch. But I'll make a separate PR so that we don't have to wait for your NFQ to be merge for this fix to be available to the rest of us. I also took the liberty making a couple of quick tweaks to your code. 
It also meant I had to change the QBasedPolicy code.

Reply:
It's better now, but not quite there yet. With the fixes you pushed, the experiment runs, but the loop runs as many times as there are steps in the episode since the trajectories sample every step; I only need to call `optimise!` once. I've applied a quick fix in 2ed5ffb but this is obviously not ideal.

It may be a good idea to implement a "sample at the end of an episode" controller in `ReinforcementLearningTrajectories.jl`, but I propose to keep that outside the scope of this PR.